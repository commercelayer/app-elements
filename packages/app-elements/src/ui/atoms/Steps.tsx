import cn from 'classnames'
import { useCallback, useMemo } from 'react'

interface Step {
  label: string
  active?: boolean
}

export interface StepsProps {
  steps: Step[]
}

export const Steps: React.FC<StepsProps> = ({ steps }) => {
  const lastActiveIndex = useMemo(
    () => steps.findLastIndex((step) => step.active === true),
    [steps]
  )

  const fixActive = useCallback(
    (step: Step, index: number): Step => {
      if (index < lastActiveIndex) {
        return {
          ...step,
          active: true
        }
      }

      return step
    },
    [lastActiveIndex]
  )

  return (
    <ul
      style={{
        backgroundPosition:
          lastActiveIndex > 0
            ? `${(lastActiveIndex / (steps.length - 1)) * 100 * -1}%`
            : ''
      }}
      className={cn(
        'flex justify-between text-xs w-full items-center rounded',
        `h-2 mb-8 bg-gradient-to-r from-gray-100 from-50% to-primary to-50% bg-[length:200%]`
      )}
    >
      {steps.map(fixActive).map((step, index) => {
        const position =
          index === 0 ? 'first' : index === steps.length - 1 ? 'last' : 'other'

        const activePosition =
          index < lastActiveIndex
            ? 'before'
            : index === lastActiveIndex
              ? 'active'
              : 'after'

        return (
          <li
            key={step.label}
            className={cn('relative', {
              'text-gray-400 font-semibold': activePosition === 'before',
              'text-black font-bold': activePosition === 'active',
              'text-gray-300 font-semibold': activePosition === 'after'
            })}
          >
            <div
              className={cn(
                'w-6 h-6 rounded-full flex justify-center items-center',
                {
                  'bg-primary': step.active === true,
                  'bg-gray-100': step.active !== true,
                  'after:bg-white after:w-3 after:h-3 after:block after:rounded-full':
                    activePosition === 'active'
                }
              )}
            />
            <div
              className={cn('absolute whitespace-nowrap mt-1', {
                'translate-x-0': position === 'first',
                'translate-x-[calc(-50%+12px)]': position === 'other',
                'right-0': position === 'last'
              })}
            >
              {step.label}
            </div>
          </li>
        )
      })}
    </ul>
  )
}

Steps.displayName = 'Steps'
