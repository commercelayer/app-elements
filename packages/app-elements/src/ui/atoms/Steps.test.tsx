import { render } from '@testing-library/react'
import { Steps } from './Steps'

describe('Steps', () => {
  it('Should be rendered', () => {
    const { container } = render(
      <Steps
        steps={[{ label: 'Step 1' }, { label: 'Step 2' }, { label: 'Step 3' }]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('Should render as many `li` as many steps are provided', () => {
    const { getByRole } = render(
      <Steps
        steps={[{ label: 'Step 1' }, { label: 'Step 2' }, { label: 'Step 3' }]}
      />
    )

    expect(getByRole('list').children.length).toEqual(3)
    expect(getByRole('list').children[0]?.textContent).toEqual('Step 1')
    expect(getByRole('list').children[1]?.textContent).toEqual('Step 2')
    expect(getByRole('list').children[2]?.textContent).toEqual('Step 3')
  })

  it('Should highlight all non "active" steps', () => {
    const { getByText } = render(
      <Steps
        steps={[{ label: 'Step 1' }, { label: 'Step 2' }, { label: 'Step 3' }]}
      />
    )

    expect(getByText('Step 1').previousSibling).toHaveClass('bg-gray-100')
    expect(getByText('Step 2').previousSibling).toHaveClass('bg-gray-100')
    expect(getByText('Step 3').previousSibling).toHaveClass('bg-gray-100')
  })

  it('Should highlight the "active" steps', () => {
    const { getByText } = render(
      <Steps
        steps={[
          { label: 'Step 1', active: true },
          { label: 'Step 2', active: true },
          { label: 'Step 3' }
        ]}
      />
    )

    expect(getByText('Step 1').previousSibling).toHaveClass('bg-primary')
    expect(getByText('Step 1').previousSibling).not.toHaveClass(
      'after:bg-white'
    )
    expect(getByText('Step 2').previousSibling).toHaveClass(
      'bg-primary after:bg-white'
    )
    expect(getByText('Step 3').previousSibling).toHaveClass('bg-gray-100')
    expect(getByText('Step 3').previousSibling).not.toHaveClass(
      'after:bg-white'
    )
  })

  it('Should mark as "active" all previous step from the latest "active"', () => {
    const { getByText } = render(
      <Steps
        steps={[
          { label: 'Step 1' },
          { label: 'Step 2', active: true },
          { label: 'Step 3' }
        ]}
      />
    )

    expect(getByText('Step 1').previousSibling).toHaveClass('bg-primary')
    expect(getByText('Step 1').previousSibling).not.toHaveClass(
      'after:bg-white'
    )
    expect(getByText('Step 2').previousSibling).toHaveClass(
      'bg-primary after:bg-white'
    )
    expect(getByText('Step 3').previousSibling).toHaveClass('bg-gray-100')
    expect(getByText('Step 3').previousSibling).not.toHaveClass(
      'after:bg-white'
    )
  })
})
