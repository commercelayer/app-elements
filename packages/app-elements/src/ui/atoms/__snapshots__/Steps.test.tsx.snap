// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Steps > Should be rendered 1`] = `
<div>
  <ul
    class="flex justify-between text-xs w-full items-center rounded h-2 mb-8 bg-gradient-to-r from-gray-100 from-50% to-primary to-50% bg-[length:200%]"
  >
    <li
      class="relative text-gray-300 font-semibold"
    >
      <div
        class="w-6 h-6 rounded-full flex justify-center items-center bg-gray-100"
      />
      <div
        class="absolute whitespace-nowrap mt-1 translate-x-0"
      >
        Step 1
      </div>
    </li>
    <li
      class="relative text-gray-300 font-semibold"
    >
      <div
        class="w-6 h-6 rounded-full flex justify-center items-center bg-gray-100"
      />
      <div
        class="absolute whitespace-nowrap mt-1 translate-x-[calc(-50%+12px)]"
      >
        Step 2
      </div>
    </li>
    <li
      class="relative text-gray-300 font-semibold"
    >
      <div
        class="w-6 h-6 rounded-full flex justify-center items-center bg-gray-100"
      />
      <div
        class="absolute whitespace-nowrap mt-1 right-0"
      >
        Step 3
      </div>
    </li>
  </ul>
</div>
`;
